---
title: "R Notebook"
output: html_notebook
---


```{r load_library, message=FALSE}
library(readr)
library(minfi)
library(dplyr)
library(shinyMethyl)
library(ggplot2)
library(limma)
library(DMRcate)
library(reshape2)
library(GenomicFeatures)
library(biomaRt)
library(biovizBase)
library(ggbio)
library(cowplot)
library(rtracklayer)
# library(TCGAbiolinks)

# load("BC123459_only_Preprocessed.RData")
# run_name <- "BC123459_only"
load("2nd_go_TM_BBRemove_Preprocessed.RData")
run_name <- "2nd_go_TM_BBRemove"
```


```{r pca_v1}
pca                 <- norm_data_f %>% getM %>% t %>% prcomp
d                   <- pca$x %>% as.data.frame %>% add_rownames("sentrixID") %>% 
                       left_join((pData(norm_data_f) %>% as.data.frame)) %>% 
                       mutate_each(funs(factor),Tiss_Conv2) %>% 
                       as.data.frame
pcv                 <- round((pca$sdev)^2 / sum(pca$sdev^2)*100, 2)

gg                  <- ggplot(d, aes(PC1,PC2)) +
                       geom_point(aes(colour = sample, shape=conversion, label=sample)) + 
                       theme_bw() +
                       ggtitle("PCA of Normalised Methylation Data") +
                       theme(axis.title.x    = element_text(size=15),
                             axis.title.y    = element_text(size=15)) +
                       xlab(label = paste0("PC (", pcv[1], "%)")) +
                       ylab(label = paste0("PC (", pcv[2], "%)")) + 
                       geom_text(aes(label=sample), nudge_x=50)#+
                       # geom_text(aes(label = Sample_Name))
print(gg)
  
png(paste0("Results_Final/PCA/PCA_",run_name,".png"),
      width  = 10.98,height = 10.98,
      units  = "in",res    = 600)
print(gg); dev.off()

# Scree Plot
# foo <- data.frame(PC_Variation = pcv, Number = 1:20)
# ggplot(foo, aes(Number, PC_Variation)) + geom_point() + geom_line() + theme_bw() + ylab("% Variation Explained") + xlab("Principle Component")
```

```{r unsupervised_clustring_tissue_type}
Mvals = getM(norm_data_f)
colnames(Mvals) = pheno_in$sample
clusters <- hclust(dist(t(Mvals)))
plot(clusters)


png(paste0("Results_Final/Clustering//HierarchicalClustering_",run_name,".png"),
      width  = 10.98,height = 10.98,
      units  = "in",res    = 600)
plot(clusters); dev.off()

# if it's not plotting in the window
# Run dev.off() in the console first
# then try again
```

```{r beta_dist}
pheno_tmp <- pData(norm_data_f) %>% as.data.frame %>% 
             mutate(Type = ifelse(grepl("normal",grade),"Normal", "Tumour"))
df        <- norm_data_f %>% 
             getBeta %>% 
             melt(value.name = "beta") %>% 
             left_join(pheno_tmp, by = c("Var2"="sentrixID")) %>% 
             filter(sample != "Blood", sample != "Brain")
             # filter(grepl("normal_ntissue|hgrade_ttissue", Tiss_Conv2),
             #        sample != "BT3010")

gg        <- ggplot(df, aes(beta, colour = sample, linetype = conversion)) +
             geom_density() + theme_bw() +
             scale_colour_brewer(palette = "Set1") +
             # scale_color_manual(values=cbPalette)
             facet_grid(. ~ Type)
print(gg)


png(paste0("Results_Final/Beta_Distributions/BetaDist_",run_name,".png"),
      width  = 10.98,height = 10.98,
      units  = "in",res    = 600)
print(gg); dev.off()


```



```{r 5HmC_DM}
model_type            <- "Paired"
pheno_in              <- pData(norm_data_f) %>% as.data.frame 
design                <- model.matrix(~0 + Tiss_Conv2, data = pheno_in)
block_in              <- pheno_in$sample
colnames(design)      <- colnames(design) %>% gsub("Tiss_Conv2","",.)
cont                  <- c("Tumour_Tissue"                  = "hgrade_ttissue_BS-hgrade_ttissue_OXBS",
                           "Normal_Tissue"                  = "normal_ntissue_BS-normal_ntissue_OXBS")
                           # "Whole_Blood"                    = "normal_Blood_BS-normal_Blood_OXBS",
                           # "Brain_Tissue"                   = "normal_Brain_BS-normal_Brain_OXBS")
cont_mat              <- makeContrasts(contrasts = cont, levels = colnames(design))
colnames(cont_mat)    <- names(cont)
cont_vec              <- cont %>% gsub("-","_Vs_",.)
mVal                  <- getM(norm_data_f)
bVal                  <- getBeta(norm_data_f)
fitm                  <- lmFit(getM(norm_data_f),    design, block = block_in, correlation = Corr$cor) %>% 
                         contrasts.fit(cont_mat) %>% eBayes
fitb                  <- lmFit(getBeta(norm_data_f), design, block = block_in, correlation = Corr$cor) %>%
                         contrasts.fit(cont_mat) %>% eBayes
contrasts             <- colnames(cont_mat)
pVal                  <- 0.05
fc                    <- 0.1
```


```{r 5HmC_DMP_Group}
model_type            <- "Grouped"
pheno_in              <- pData(norm_data_f) %>% as.data.frame
design                <- model.matrix(~0 + Tiss_Conv2, data = pheno_in)
colnames(design)      <- colnames(design) %>% gsub("Tiss_Conv2","",.)
cont                  <- c("Tumour-Normal_BS"               = "hgrade_ttissue_BS-normal_ntissue_BS",
                           "Tumour-Normal_OXBS"             = "hgrade_ttissue_OXBS-normal_ntissue_OXBS")
cont_mat              <- makeContrasts(contrasts = cont, levels = colnames(design))
colnames(cont_mat)    <- names(cont)
cont_vec              <- cont %>% gsub("-","_Vs_",.)
fitm                  <- lmFit(getM(norm_data_f),    design) %>%
                         contrasts.fit(cont_mat) %>% eBayes
fitb                  <- lmFit(getBeta(norm_data_f), design) %>%
                         contrasts.fit(cont_mat) %>% eBayes
contrasts             <- colnames(cont_mat)
pVal                  <- 0.05
fc                    <- 0.1
```



```{r Results, cache=FALSE, message=FALSE, warning=FALSE, error=FALSE, results='hide'}
tt_out  <- list()
ttu_out <- list()
for(i in names(cont_vec)) {
  ttm                  <- topTable(fitm, coef = i, number = Inf, p.value = pVal, adjust.method = "BH") %>% 
                          as.data.frame %>% add_rownames("CpG")
  ttum                 <- topTable(fitm, coef = i, number = Inf, adjust.method = "BH") %>% 
                          as.data.frame %>% add_rownames("CpG")
  if(nrow(ttm) > 0) {
    ttm                  <- ttm %>% dplyr::select(CpG,`adj.P.Val`)
    ttb                  <- topTable(fitb, coef = i, number = Inf) %>% 
                            as.data.frame %>% add_rownames("CpG") %>% dplyr::select(CpG,Delta_Beta = logFC)
    tt                   <- ttm %>% left_join(ttb)  %>% 
                                    left_join(annotation, by = c("CpG" = "Name")) %>% 
                                    filter(abs(Delta_Beta) > fc) %>% arrange(desc(abs(Delta_Beta)))
    
    ttum                 <- ttum %>% dplyr::select(CpG,`adj.P.Val`)
    ttub                 <- topTable(fitb, coef = i, number = Inf) %>% 
                            as.data.frame %>% add_rownames("CpG") %>% dplyr::select(CpG,Delta_Beta = logFC)
    ttu                  <- ttum %>% left_join(ttub)  %>% 
                                     left_join(annotation, by = c("CpG" = "Name"))
    
    tt %>% print
    tt_out[[i]]          <- tt
    ttu_out[[i]]         <- ttu
    # write_tsv(tt, path = paste0("Results_Final/5HmC_DMP/BS_-_OXBS_DMP_",model_type,"_",i,"_",run_name,".txt"))
    # write_tsv(tt, path = paste0("Results_Final/5HmC_DMP/DMP_",model_type,"_",i,"_",run_name,".txt"))
  } else {
    paste0(i, ": No DM Results") %>% print
  }
}
```


```{r check}
# foo <- read_csv("noBc6/1 Normal_BS-OXBS_noBC6.csv") %>% filter(Delta_Beta > 0)
# foo
# baa <- tt_out[[1]]
# baa
# intersect(foo$CpG, baa$CpG) %>% length
```


```{r DMR_Detection}
# design_DMR                <- model.matrix(~sample + Tiss_Conv, data = pheno_in)
# colnames(design_DMR)      <- colnames(design_DMR) %>% gsub("Tiss_Conv","",.)

DMR_List_Out <- list()
for(i in colnames(cont_mat)) {
  myannotation       <- cpg.annotate(datatype      = "array",
                                     what          = "M",
                                     arraytype     = "450K",
                                     object        = mVal, 
                                     analysis.type = "differential", 
                                     design        = design, 
                                     coef          = i,
                                     contrasts     = T, 
                                     cont.matrix   = cont_mat,
                                     fdr           = 0.01)
  # For Paired DMR Modelling 
    # source("Cpg_Annotate2.R")
    # myannotation       <- cpg.annotate2(datatype      = "array",
    #                                  what          = "M",
    #                                  arraytype     = "450K",
    #                                  object        = mVal, 
    #                                  analysis.type = "differential", 
    #                                  design        = design, 
    #                                  coef          = i,
    #                                  contrasts     = T, 
    #                                  cont.matrix   = cont_mat,
    #                                  corr_coef     = Corr$cor,
    #                                  blocking_factor = block_in,
    #                                  fdr           = 0.01)
  
  dmrcoutput         <- dmrcate(myannotation) 
                      
  dmrresults         <- dmrcoutput$results %>% 
                        as.data.frame %>% 
                        mutate(betaAfc = abs(meanbetafc)) %>% 
                        filter(betaAfc > 0.1) %>% 
                        arrange(desc(no.cpgs))
  
  dmr_genes_tmp <- sapply(dmrresults$coord, function(x) {
    dmr_coords           <- x
    dmr_chr              <- strsplit(dmr_coords, ':')[[1]][1]
    dmr_start            <- as.numeric(strsplit(dmr_coords, '-|:')[[1]][2])
    dmr_end              <- as.numeric(strsplit(dmr_coords, '-')[[1]][2])
    dmr_anno             <- annotation %>% 
                            filter(chr == dmr_chr,
                                   pos >= (dmr_start),
                                   pos <= (dmr_end))
    dmr_genes            <- dmr_anno$UCSC_RefGene_Name %>% unique %>% .[. != ""]
    dmr_enhancer         <- dmr_anno$Enhancer %>% unique %>% .[. != ""]
    if(length(dmr_genes) == 0 & length(dmr_enhancer) == 0) {
      return(c("",""))
    } else {
      dmr_genes_tmp        <- sapply(dmr_genes, function(y) {
        gene_split_tmp       <- strsplit(y,";")[[1]] %>% unique
        return(gene_split_tmp)
      })
      dmr_enhancers_tmp    <- sapply(dmr_enhancer, function(y) {
        enhancer_split_tmp   <- strsplit(y,";")[[1]] %>% unique
        return(enhancer_split_tmp)
      })
      
      dmr_genes            <- dmr_genes_tmp %>% unlist %>% unique %>% paste(., collapse = ";")
      dmr_enhancers        <- dmr_enhancers_tmp %>% unlist %>% unique %>% paste(., collapse = ";")
      return(c(dmr_genes,dmr_enhancers))
      # return(dmr_genes)
    }
  })
  gene_enhancer_tmp <- dmr_genes_tmp %>% unlist
  
  
  dmrresults$Genes      <- gene_enhancer_tmp[seq(1,length(gene_enhancer_tmp),2)]
  dmrresults$Enhancers  <- gene_enhancer_tmp[seq(2,length(gene_enhancer_tmp),2)]
  DMR_List_Out[[i]]     <- dmrresults
  
  write_csv(dmrresults, path = paste0("Results_Final/5HmC_DMR/5HmC_DMR_",i,"_",run_name,"_",model_type,".csv"))
  print(dmrresults)
}

# DMR_List_Out[[1]] %>% .$meanbetafc %>% .[. > 0] %>% length
# DMR_List_Out[[1]] %>% .$meanbetafc %>% .[. < 0] %>% length
# DMR_List_Out[[2]] %>% .$meanbetafc %>% .[. > 0] %>% length
# DMR_List_Out[[2]] %>% .$meanbetafc %>% .[. < 0] %>% length

# Split this script into two.
#   - One for 5HmC
#   - One for Tumour Normal Analysis
# Make two difference output directories
# Parse DMR regions, get CpG names back for each region, then perform upset plot. - Done 
# Redo the DMR to gene expression relationships
```


```{r upset_and_barchart}
library(UpSetR)

df_upset          <- data.frame(CpG = "")
for(i in names(tt_out)) {
  tmpdf              <- data.frame(CpG = tt_out[[i]]$CpG, DM = 1)
  colnames(tmpdf)[2] <- paste0("5HmC_Positions_", i)
  
  df_upset           <- df_upset %>% 
                        full_join(tmpdf)
  
  # Write out sets
  oxbs_bs_intersection <- data.frame(Gene_Symbol = intersect(unique(dmr_to_gene_out[["OXBS"]][[i]]$Gene),
                                                             unique(dmr_to_gene_out[["BS"]][[i]]$Gene)))
  oxbs_only            <- data.frame(Gene_Symbol = setdiff(unique(dmr_to_gene_out[["OXBS"]][[i]]$Gene),
                                                           unique(dmr_to_gene_out[["BS"]][[i]]$Gene)))
  bs_only              <- data.frame(Gene_Symbol = setdiff(unique(dmr_to_gene_out[["BS"]][[i]]$Gene),
                                                           unique(dmr_to_gene_out[["OXBS"]][[i]]$Gene)))
  write_csv(oxbs_bs_intersection, path = paste0("Sets/OXBS_BS_Intersect_",i,".csv"))
  write_csv(oxbs_only, path = paste0("Sets/OXBS_Only_",i,".csv"))
  write_csv(bs_only, path = paste0("Sets/BS_Only_",i,".csv"))
}
df_upset[is.na(df_upset)] <- 0


png(paste0("Results_Final/Upset/UpSetPlot_","5HmC_Tumour_Normal",".png"),
      width  = 13,height = 7.5,
      units  = "in",res    = 600)
upset(df_upset, nsets = ncol(df_upset)); dev.off()

TN_intersect <- data.frame(CpG = intersect(tt_out$Tumour_Tissue$CpG,tt_out$Normal_Tissue$CpG)) %>% 
                left_join(annotation, by = c("CpG" = "Name"))
T_Only       <- data.frame(CpG = setdiff(tt_out$Tumour_Tissue$CpG,tt_out$Normal_Tissue$CpG)) %>% 
                left_join(annotation, by = c("CpG" = "Name"))
N_Only       <- data.frame(CpG = setdiff(tt_out$Normal_Tissue$CpG,tt_out$Tumour_Tissue$CpG)) %>% 
                left_join(annotation, by = c("CpG" = "Name"))
write_csv(TN_intersect, path = "Results_Final/Upset/Sets/Tumour_Normal_5HmC_Intersect.csv")
write_csv(T_Only, path = "Results_Final/Upset/Sets/Tumour_Only_5HmC.csv")
write_csv(N_Only, path = "Results_Final/Upset/Sets/Normal_Only_5HmC.csv")


# for(i in names(tt_out)) {
#   df_upset_tmp <- df_upset %>% dplyr::select(CpG, ends_with(i))
#   png(paste0("Sets/UpSetPlot_CellTypes_BS-OXBS_",i,".png"),
#       width  = 13,height = 7.5,
#       units  = "in",res    = 600)
#   upset(df_upset_tmp, nsets = 2); dev.off()
# }

df_bar <- c()
for(i in names(tt_out)) {
  tmp_df     <- df_upset %>% dplyr::select(ends_with(i))
  tmp_sum    <- sum(tmp_df[[1]])
  tmp_df_out <- data.frame(Tissue = i, 
                           Class  = c("Sum"),
                           Counts = c(tmp_sum))
  df_bar     <- rbind(df_bar, tmp_df_out)
}
df_bar_in <- df_bar
gg        <- ggplot(df_bar_in, aes(Tissue,Counts)) +
             geom_bar(stat = "identity", position = "dodge") + 
             theme_bw() + 
             geom_text(aes(label=Counts), vjust=0) +
             theme(axis.text.x=element_text(angle = -90, hjust = 0)) +
             ggtitle("The Number of 5HmC Sites found in each Tissue Group")

png("Results_Final/Upset/5HmC_TumourNormal_BarChart_Total.png",
      width  = 7.5,height = 7.5,
      units  = "in",res    = 600)
print(gg); dev.off()


```



```{r DMR_Region_to_CpGs}
DMR_to_CpG <- list()
for(i in 1:length(DMR_List_Out)) {
  DMR_Temp     <- DMR_List_Out[[i]] #%>% filter(meanbetafc > 0)
  chr_tmp      <- DMR_Temp$coord %>% gsub(":.*$", "",.)
  coords_tmp   <- DMR_Temp$coord %>% gsub( "^.*:", "",.)
  start_tmp    <- coords_tmp %>% gsub( "-.*$", "",.)
  end_tmp      <- coords_tmp %>% gsub( "^.*-", "",.)
  
  df_tmp       <- data.frame(CHR = chr_tmp, START = start_tmp, END = end_tmp) %>% 
                  mutate_each(funs(as.character), START, END) %>% 
                  mutate_each(funs(as.numeric), START, END)
  df_tmp$meanbetafc <- DMR_Temp$meanbetafc[i]
  df_out_tmp        <- c()
  for(j in 1:nrow(df_tmp)) {
    anno_tmp <- annotation %>% filter(chr == df_tmp$CHR[j],
                                      pos >= df_tmp$START[j],
                                      pos <= df_tmp$END[j])
    anno_tmp$meanbetafc <- df_tmp$meanbetafc[i]
    df_out_tmp <- rbind(df_out_tmp, anno_tmp)
  }
  DMR_to_CpG[[names(DMR_List_Out)[i]]] <- df_out_tmp
}

## ^ This takes DMR to CpG sites, but preserves mean delta beta value. 
## to look for 
## DMR - ChrZ:X:Y  -> This has a gene symbol (nearest gene, or many genes, if it overlaps with many genes)
## Gene Expression -> Whole expression of a gene.
## DMR2CpG         -> lots of CpGs that have their attached delta beta of their parent DMR
## How to marry up DMR2CpG and Gene Expression??
## Previously based on gene symbols. 
## To Match:
## DMR2CpG, convert to irange/ genomic range. Liftover to hg38. 
## Then for each unique delta beta value, take the first CpG bp pos, and the last CpG bp pos. 
## Make that into an iRanges/ Genomic ranges object
## Calculate Overlaps



df_upset          <- data.frame(CpG = "")
for(i in names(DMR_to_CpG)) {
  tmpdf              <- data.frame(CpG = DMR_to_CpG[[i]]$Name, DM = 1)
  colnames(tmpdf)[2] <- paste0("5HmC_Positions_", i)
  
  df_upset           <- df_upset %>% 
                        full_join(tmpdf)
}
df_upset[is.na(df_upset)] <- 0

dec_exp_inc_methy <- df_out %>% filter(logFC < 0, Delta_B > 0) %>% distinct %>% arrange(desc(CpG_Num))
inc_exp_dec_methy <- df_out %>% filter(logFC > 0, Delta_B < 0) %>% distinct
weird_ones        <- df_out %>% filter((logFC > 0 & Delta_B > 0) | (logFC < 0 & Delta_B < 0)) %>% distinct
dec_exp_inc_methy; inc_exp_dec_methy; weird_ones

list_tmp <- list()
list_tmp[["dec_exp_inc_methy"]] <- dec_exp_inc_methy
list_tmp[["inc_exp_dec_methy"]] <- inc_exp_dec_methy
list_tmp[["consis_high"]]       <- weird_ones %>% filter(logFC > 0)
list_tmp[["consis_low"]]        <- weird_ones %>% filter(logFC < 0)
dmr_to_gene_out[[runs_in[z]]]   <- list_tmp



# Write out sets
# OXBS_BS_intersection <- data.frame(CpG = intersect(DMR_to_CpG[[1]]$Name, DMR_to_CpG[[2]]$Name)) %>% 
#                         left_join(annotation, by = c("CpG" = "Name"))     
# BS_only              <- data.frame(CpG = setdiff(DMR_to_CpG[[1]]$Name, DMR_to_CpG[[2]]$Name)) %>% 
#                         left_join(annotation, by = c("CpG" = "Name"))  
# OXBS_only            <- data.frame(CpG = setdiff(DMR_to_CpG[[2]]$Name, DMR_to_CpG[[1]]$Name)) %>% 
#                         left_join(annotation, by = c("CpG" = "Name")) 
# write_csv(OXBS_BS_intersection, path = paste0("Results_Final/DMR/Upset/OXBS_BS_Intersect_TN.csv"))
# write_csv(BS_only, path = paste0("Results_Final/DMR/Upset/BS_Only_TN.csv"))
# write_csv(OXBS_only, path = paste0("Results_Final/DMR/Upset/OXBS_Only_TN.csv"))
# 

png(paste0("Results_Final/DMR/UpSetPlot_","Tmour_Normal_DMR",".png"),
      width  = 13,height = 7.5,
      units  = "in",res    = 600)
upset(df_upset, nsets = ncol(df_upset)); dev.off()
```



```{r liftover_hg19_to_hg38}
chain.file.in <- "~/Downloads/hg19ToHg38.over.chain"
chain.in      <- import.chain(chain.file.in)
# str(chain.in)

dmr_list_lo   <- list()
for(i in names(DMR_to_CpG)) {
  dmr_list_in   <- DMR_to_CpG[[i]]
  hg19.range    <- GRanges(seqnames = Rle(dmr_list_in$chr), 
                           ranges = IRanges(start = dmr_list_in$pos, 
                                            end   = dmr_list_in$pos,
                                            names = dmr_list_in$Name), strand = dmr_list_in$strand)
  hg38.range    <- liftOver(hg19.range, chain.in)
  hg38.out      <- as.data.frame(unlist(hg38.range)) %>% add_rownames("CpG") %>% 
                   dplyr::select(CpG, chr = seqnames, pos = start, strand)
  dmr.out       <- dmr_list_in %>% dplyr::select(-chr,-pos,-strand) %>% 
                   right_join(hg38.out, by = c("Name"="CpG")) %>% 
                   dplyr::select(Name, chr, pos, strand, everything())
  dmr_list_lo[[i]] <- dmr.out
}

                
```


```{r dmr_p2gene}
dmr_to_gene_out <- list()
runs_in         <- c("BS","OXBS")
buffer          <- 1e6
for(z in 1:length(dmr_list_lo)) {
  print(names(dmr_list_lo)[z])
  
  dmrresults <- dmr_list_lo[[z]]
  df_out     <- c()
  
  for(x in 1:nrow(dmrresults)) {
    dmr_tmp  <- dmrresults[x,]
    res_tmp  <- res %>% filter(chromosome_name == gsub("chr","",dmr_tmp$chr),
                               (start_position - buffer) < dmr_tmp$pos,
                               (end_position + buffer) > dmr_tmp$pos) %>% 
                mutate(cpg_diff = abs(end_position - dmr_tmp$pos)) %>% 
                arrange(cpg_diff)
                
    if(nrow(res_tmp) == 0) { next; }
    
    df_tmp <- data.frame(DMR      = dmr_tmp$Name,
                         Delta_B  = dmr_tmp$meanbetafc,
                         Gene     = res_tmp$hgnc_symbol[1],
                         logFC    = res_tmp$log2FoldChange[1])
    df_out <- rbind(df_out,df_tmp)
  }
  
  dec_exp_inc_methy <- df_out %>% filter(logFC < 0, Delta_B > 0) %>% distinct
  inc_exp_dec_methy <- df_out %>% filter(logFC > 0, Delta_B < 0) %>% distinct
  weird_ones        <- df_out %>% filter((logFC > 0 & Delta_B > 0) | (logFC < 0 & Delta_B < 0)) %>% distinct
  dec_exp_inc_methy; inc_exp_dec_methy; weird_ones
  
  list_tmp <- list()
  list_tmp[["dec_exp_inc_methy"]] <- dec_exp_inc_methy
  list_tmp[["inc_exp_dec_methy"]] <- inc_exp_dec_methy
  list_tmp[["consis_high"]]       <- weird_ones %>% filter(logFC > 0)
  list_tmp[["consis_low"]]        <- weird_ones %>% filter(logFC < 0)
  dmr_to_gene_out[[runs_in[z]]]   <- list_tmp
}

write_csv(dmr_to_gene_out[["OXBS"]][["dec_exp_inc_methy"]], path = paste0("Results_Final/DMR_Gene_Correlations/dec_exp_inc_methy_OXBS_",run_name,".csv"))
write_csv(dmr_to_gene_out[["OXBS"]][["inc_exp_dec_methy"]], path = paste0("Results_Final/DMR_Gene_Correlations/inc_exp_dec_methy_OXBS_",run_name,".csv"))
write_csv(dmr_to_gene_out[["OXBS"]][["consis_high"]], path = paste0("Results_Final/DMR_Gene_Correlations/consis_high_OXBS_",run_name,".csv"))
write_csv(dmr_to_gene_out[["OXBS"]][["consis_low"]], path = paste0("Results_Final/DMR_Gene_Correlations/consis_low_OXBS_",run_name,".csv"))

write_csv(dmr_to_gene_out[["BS"]][["dec_exp_inc_methy"]], path = paste0("Results_Final/DMR_Gene_Correlations/dec_exp_inc_methy_BS_",run_name,".csv"))
write_csv(dmr_to_gene_out[["BS"]][["inc_exp_dec_methy"]], path = paste0("Results_Final/DMR_Gene_Correlations/inc_exp_dec_methy_BS_",run_name,".csv"))
write_csv(dmr_to_gene_out[["BS"]][["consis_high"]], path = paste0("Results_Final/DMR_Gene_Correlations/consis_high_BS_",run_name,".csv"))
write_csv(dmr_to_gene_out[["BS"]][["consis_low"]], path = paste0("Results_Final/DMR_Gene_Correlations/consis_low_BS_",run_name,".csv"))

# Next - Upset plots for above code. and make sure gene numbers are correct on plots 
# Comparing the CpGs-gene relationships 

df_upset          <- data.frame(CpG_Gene_Map = "")
for(i in names(dmr_to_gene_out)) {
  for(j in names(dmr_to_gene_out[[i]])) {
    tmp_in    <- dmr_to_gene_out[[i]][[j]] %>% 
                 mutate(TMP = paste0(Delta_B,logFC)) %>% 
                 distinct(TMP, .keep_all = TRUE)
    tmp_df    <- data.frame(CpG_Gene_Map = paste0(tmp_in$DMR, "_", tmp_in$Gene), 
                            DM = 1)
    colnames(tmp_df)[2] <- paste0(i,"_",j)
    df_upset  <- df_upset %>% full_join(tmp_df)
  }
}
df_upset[is.na(df_upset)] <- 0
df_upset <- df_upset[-1,] %>% filter(!grepl("^_", CpG_Gene_Map)) %>% 
            filter(!grepl("_$", CpG_Gene_Map)) 
  
upset(df_upset, sets = colnames(df_upset)[-1])

png(paste0("Results_Final/DMR/UpSetPlot_","Tmour_Normal_DMR",".png"),
      width  = 13,height = 7.5,
      units  = "in",res    = 600)
upset(df_upset, nsets = (ncol(df_upset)+10)); dev.off()



# Need TCGA and TM XY compare
# Spliting 5hmc and mc result into two folders and Rmardown files. 



```







```{r UpSetPlotting}
library(UpSetR)

df_upset        <- data.frame(Gene_Symbol = "")
vars_to_process <- c("dec_exp_inc_methy", "inc_exp_dec_methy", "consis_high", "consis_low")

for(i in vars_to_process) {
  oxbs_tmpdf <- data.frame(Gene_Symbol = unique(dmr_to_gene_out[["OXBS"]][[i]]$Gene), OXBS = 1)
  bs_tmpdf   <- data.frame(Gene_Symbol = unique(dmr_to_gene_out[["BS"]][[i]]$Gene), BS = 1)
  
  colnames(oxbs_tmpdf)[2] <- paste0("OXBS_", i)
  colnames(bs_tmpdf)[2]   <- paste0("BS_", i)
  
  df_upset   <- df_upset %>% 
                full_join(oxbs_tmpdf) %>% 
                full_join(bs_tmpdf) 
  
  # Write out sets
  oxbs_bs_intersection <- data.frame(Gene_Symbol = intersect(unique(dmr_to_gene_out[["OXBS"]][[i]]$Gene),
                                                           unique(dmr_to_gene_out[["BS"]][[i]]$Gene)))
  oxbs_only            <- data.frame(Gene_Symbol = setdiff(unique(dmr_to_gene_out[["OXBS"]][[i]]$Gene),
                                                           unique(dmr_to_gene_out[["BS"]][[i]]$Gene)))
  bs_only              <- data.frame(Gene_Symbol = setdiff(unique(dmr_to_gene_out[["BS"]][[i]]$Gene),
                                                           unique(dmr_to_gene_out[["OXBS"]][[i]]$Gene)))
  # write_csv(oxbs_bs_intersection, path = paste0("Sets/OXBS_BS_Intersect_",i,".csv"))
  # write_csv(oxbs_only, path = paste0("Sets/OXBS_Only_",i,".csv"))
  # write_csv(bs_only, path = paste0("Sets/BS_Only_",i,".csv"))
}
df_upset[is.na(df_upset)] <- 0
# write_csv(df_upset, path = "Sets/UpSet_Table.csv")
# replace_na(list(OXBS = 0, BS = 0)) 
# df_upset   <- oxbs_tmpdf %>% full_join(bs_tmpdf) %>% replace_na(list(OXBS = 0, BS = 0)) 


png("Sets/UpSetPlot.png",
      width  = 13,height = 7.5,
      units  = "in",res    = 600)
upset(df_upset, nsets = 8); dev.off()

print(vars_to_process)
for(i in vars_to_process) {
  df_upset_tmp <- df_upset %>% dplyr::select(Gene_Symbol, ends_with(i))
  # png(paste0("Sets/UpSetPlot_",i,".png"),
  #     width  = 13,height = 7.5,
  #     units  = "in",res    = 600)
  upset(df_upset_tmp, nsets = 2)#; dev.off()
}

```





