---
title: "analysis 1"
author: "Alrayan"
date: "08/06/2017"
output: html_document
---

```{r load_library, message=FALSE}
library(readr)
library(minfi)
library(dplyr)
library(shinyMethyl)
library(ggplot2)
library(limma)
library(DMRcate)
library(reshape2)
library(GenomicFeatures)
library(biomaRt)
library(biovizBase)
library(ggbio)
library(cowplot)
# library(TCGAbiolinks)
```

```{r pheno_load}

# experiment_in <- "1st_go_TM" #Other Option= 2nd_go_TM
# experiment_in <- "2nd_go_TM" #Other Option= 
experiment_in <- "2nd_go_TM_BBRemove"


if(experiment_in == "1st_go_TM") {
  pheno_filter <- "BT3010|BT3104|NHU86|NHU102|EJ|RT112"
} else if(experiment_in == "2nd_go_TM") {
  pheno_filter <- "Blood|Brain|NT|BC"
} else if(experiment_in == "2nd_go_TM_BBRemove") {
  pheno_filter <- "NT|BC"
} else {
  message("Invalid Argument")
}

pheno = read_tsv("Analysis_Final_Structure/Phenotype_Data/all.array1.txt") %>% as.data.frame %>%
mutate(Basename = paste0("Analysis_Final_Structure/Data/raw_idat/",sentrixID),
       grade = ifelse(grepl("hgg",grade),"hgrade",grade), 
       tissue_type = ifelse(tissue_type == "ctissue", sample, tissue_type), 
       Tiss_Conv2 = paste0(grade,"_",tissue_type, "_", conversion)) %>% 
       filter(sample != "PC6") %>% 
       mutate(sample = ifelse(grepl("PC",sample),gsub("PC","BC",sample), sample)) %>% 
       filter(grepl(pheno_filter, sample))
      
        # sample      != "BT3010",
        # sample      != "BT3104",
        # sample      != "Blood",
        # sample      != "Brain",
        # sample      != "EJ",
        # sample      != "RT112",
        # tissue_type != "pline")
       
       
# pheno$Tiss_Conv2 %>% table
# write_tsv(pheno, path="pheno_file_in_allTM_noBC6.txt")
pheno

read_tsv("Analysis_Final_Structure/Phenotype_Data/all.array1.txt") %>% as.data.frame %>%
    mutate(Basename = paste0("Analysis_Final_Structure/Data/raw_idat/",sentrixID),
           grade = ifelse(grepl("hgg",grade),"hgrade",grade), 
           tissue_type = ifelse(tissue_type == "ctissue", sample, tissue_type), 
           Tiss_Conv2 = paste0(grade,"_",tissue_type, "_", conversion)) %>% 
    filter(sample != "PC6") %>% 
    mutate(sample = ifelse(grepl("PC",sample),gsub("PC","BC",sample), sample)) %>% .$Tiss_Conv2 %>% table()
```

```{r reading_idat}
raw_data <- read.metharray.exp(targets = pheno)
raw_data
det_pval                    <- raw_data %>% detectionP(type = "m+u")
det_pval_fail               <- det_pval > 0.01
det_pval_remove             <- names(which(rowMeans(det_pval_fail) > 0.5, TRUE))
```

```{r QC}
summary <- shinySummarize(raw_data)
runShinyMethyl(summary)
```

```{r normalisation_data}
set.seed(123)
norm_data  <- minfi::preprocessFunnorm(raw_data)
norm_data
```

```{r probe_filter}
Mvals               <- norm_data %>% getM
Bvals               <- norm_data %>% getBeta

mutimappers         <- read_tsv("./Analysis_Final_Structure/Reference/HumanMethylation450_15017482_v.1.1_hg19_bowtie_multimap.txt", 
                                col_types=cols(), col_names = F) %>% 
                       .$X1 %>% unique
crossreact          <- read_csv("./Analysis_Final_Structure/Reference/48639-non-specific-probes-Illumina450k.csv", 
                                col_types=cols()) %>% 
                       .$TargetID %>% unique
cpg_remove_mmcr     <- c(mutimappers,crossreact) %>% unique

remove              <- which(abs(Mvals)==Inf, arr.ind=TRUE)[,1]
beta_thres          <- (Bvals > 0.8 | Bvals < 0.2) %>% 
                       apply(., 1, all) %>% 
                       .[. == T] %>% 
                       names

remove              <- c(remove,match(unique(c(det_pval_remove,beta_thres,cpg_remove_mmcr)),
                                      rownames(norm_data))) %>% 
                       unique %>% na.omit

if(length(remove) > 0) {
  norm_data_f       <- norm_data[-remove,]
}else{
  norm_data_f       <- norm_data
}
# norm_data_f         <- norm_data_f[-(match(det_pval_remove,
#                                            rownames(norm_data_f)) %>% na.omit),]

# manifest            <- raw_data %>% getManifest
norm_data_f         <- norm_data_f %>% addSnpInfo
# snps_filtered       <- norm_data_f@rowRanges %>% 
#                        as.data.frame %>% add_rownames("CpG") %>% 
#                        filter(Probe_maf > 0.05 | CpG_maf > 0.05 | SBE_maf > 0.05) %>% 
#                        .$CpG %>% as.vector

norm_data_f         <- norm_data_f %>% dropLociWithSnps(., snps = c("SBE","CpG","Probe"), maf  = 0.05)
annotation          <- norm_data_f %>% getAnnotation %>% as.data.frame
# xy_filtered         <- annotation %>% add_rownames("CpG") %>%
#                        filter(chr == "chrX" | chr == "chrY") %>%
#                        .$CpG %>% as.vector

norm_data_f         <- norm_data_f[-grep("chrX|chrY", annotation$chr),]
annotation          <- norm_data_f %>% getAnnotation %>% as.data.frame

rm(Mvals,Bvals,raw_idat,det_pval,beta_thres); gc()
# Mvals               <- norm_data_f %>% getM
# Bvals               <- norm_data_f %>% getBeta
norm_data_f
```


```{r save_preprocessing}
pheno_in              <- pData(norm_data_f) %>% as.data.frame 
design                <- model.matrix(~0 + Tiss_Conv2, data = pheno_in)
block_in              <- pheno_in$sample
colnames(design)      <- colnames(design) %>% gsub("Tiss_Conv2","",.)
Corr                  <- duplicateCorrelation(getM(norm_data_f), design, block = block_in)

save(Corr, design, norm_data_f, pheno_in, annotation, file = paste0(experiment_in,"_Preprocessed.RData"), compress = T)
```




```{r pca_v1}
pca                 <- norm_data_f %>% getM %>% t %>% prcomp
d                   <- pca$x %>% as.data.frame %>% add_rownames("sentrixID") %>% 
                       left_join((pData(norm_data_f) %>% as.data.frame)) %>% 
                       mutate_each(funs(factor),Tiss_Conv2) %>% 
                       as.data.frame
pcv                 <- round((pca$sdev)^2 / sum(pca$sdev^2)*100, 2)

gg                  <- ggplot(d, aes(PC1,PC2)) +
                       geom_point(aes(colour = sample, shape=conversion, label=sample)) + 
                       theme_bw() +
                       ggtitle("PCA of Normalised Methylation Data") +
                       theme(axis.title.x    = element_text(size=15),
                             axis.title.y    = element_text(size=15)) +
                       xlab(label = paste0("PC (", pcv[1], "%)")) +
                       ylab(label = paste0("PC (", pcv[2], "%)")) + 
                       geom_text(aes(label=sample), nudge_x=50)#+
                       # geom_text(aes(label = Sample_Name))
print(gg)
  
# png(paste0("../Analysis/Healthy_T_Cells/","PCA.png"),
#       width  = 10.98,height = 10.98,
#       units  = "in",res    = 600)
# print(gg); dev.off()

# Scree Plot
# foo <- data.frame(PC_Variation = pcv, Number = 1:20)
# ggplot(foo, aes(Number, PC_Variation)) + geom_point() + geom_line() + theme_bw() + ylab("% Variation Explained") + xlab("Principle Component")
```

```{r unsupervised_clustring_tissue_type}
Mvals = getM(norm_data_f)
colnames(Mvals) = pheno$sample
clusters <- hclust(dist(t(Mvals)))
plot(clusters)

# if it's not plotting in the window
# Run dev.off() in the console first
# then try again
```

```{r beta_dist}
pheno_tmp <- pData(norm_data_f) %>% as.data.frame %>% 
             mutate(Type = ifelse(grepl("normal",grade),"Normal", "Tumour"))
df        <- norm_data %>% 
             getBeta %>% 
             melt(value.name = "beta") %>% 
             left_join(pheno_tmp, by = c("Var2"="sentrixID")) %>% 
             filter(grepl("normal_ntissue|hgrade_ttissue", Tiss_Conv2),
                    sample != "BT3010")

gg        <- ggplot(df, aes(beta, colour = sample, linetype = conversion)) +
             geom_density() + theme_bw() +
             scale_colour_brewer(palette = "Set1") +
             # scale_color_manual(values=cbPalette)
             facet_grid(. ~ Type)
print(gg)
```



```{r scatter_plotting}
x_in      <- "normal_pline"
y_in      <- "normal_ntissue"
conv_in   <- "BS"

# df_tmp    <- norm_data_f %>% 
#              getBeta %>% 
#              melt(value.name = "beta") %>% 
#              left_join(as.data.frame(pData(norm_data_f)), by = c("Var2"="sentrixID")) %>% 
#              filter(grepl(paste(x_in,y_in,sep = "|"), Tiss_Conv2))
# 
# df_x      <- df_tmp %>% filter(Tiss_Conv2 == paste0(x_in,"_",conv_in)) %>% 
#              group_by(Var1) %>% 
#              summarise(Mean_Beta = mean(beta)) %>% 
#              ungroup
# df_y      <- df_tmp %>% filter(Tiss_Conv2 == paste0(x_in,"_",conv_in)) %>% 
#              group_by(Var1) %>% 
#              summarise(Mean_Beta = mean(beta)) %>% 
#              ungroup
# 
# df        <- data.frame(CpG    = unique(df_tmp$Var1),
#                         x_beta = df_x$Mean_Beta,
#                         y_beta = df_y$Mean_Beta)
# gg        <- ggplot(df, aes(x_beta, y_beta)) + geom_point() + theme_bw() +
#              geom_smooth(method=lm,se=FALSE)

# For the normal_pline DM sites, take the delta beta
# For the normal_ntissue DM sites, take the delta beta values
# Plot XY against each other. 

cpgs_in     <- unique( c((ttu_out[[3]] %>% filter(adj.P.Val < 0.05, Delta_Beta > 0.1) %>% .$CpG),
                         (ttu_out[[2]] %>% filter(adj.P.Val < 0.05, Delta_Beta > 0.1) %>% .$CpG)))
ttu_pline   <- ttu_out[[3]] %>% filter(CpG %in% cpgs_in) %>% dplyr::select(CpG, DeltaBeta_X = Delta_Beta)
ttu_ntissue <- ttu_out[[2]] %>% filter(CpG %in% cpgs_in) %>% dplyr::select(CpG, DeltaBeta_Y = Delta_Beta)
df_in       <- ttu_pline %>% full_join(ttu_ntissue, by = "CpG") %>% 
               mutate(Sig = ifelse(DeltaBeta_X > 0.1 & DeltaBeta_Y > 0.1, "5HmC", ""))
               # filter(DeltaBeta_X > 0.1, DeltaBeta_Y > 0.1)
# df_in[is.na(df_in)] <- 0
gg          <- ggplot(df_in, aes(DeltaBeta_X, DeltaBeta_Y, colour = Sig)) + 
               geom_point(size = 0.5) + theme_bw() +
               # geom_smooth(method=lm,se=F) +
               geom_vline(xintercept = 0) + 
               geom_hline(yintercept = 0) + 
               xlab(names(ttu_out)[3]) +
               ylab(names(ttu_out)[2]) +
               xlim(c(-1,1)) + ylim(c(-1,1)) +
               geom_hline(yintercept = c(-0.1,0.1), colour = "red") +
               geom_vline(xintercept = c(-0.1,0.1), colour = "red") +
               ggtitle("XY Scatter Plot of Delta Beta Values\nBetween BS and OXBS Sample Preps\nInferring 5HmC Sites")

png(paste0("XYScatter/XYScatter_PrimaryCL_NormTiss.png"),
      width  = 10.98,height = 10.98,
      units  = "in",res    = 600)
print(gg); dev.off()

df_in_csv <- df_in %>% 
             mutate(Class = ifelse(DeltaBeta_X > 0.1 & DeltaBeta_Y > 0.1, "Common_5HmC", ""),
                    Class = ifelse(DeltaBeta_X > 0.1 & DeltaBeta_Y < 0.1, "PrimaryCL_5HmC_Only", Class),
                    Class = ifelse(DeltaBeta_X < 0.1 & DeltaBeta_Y > 0.1, "NormTiss_5HmC_Only", Class)) %>% 
             dplyr::select(CpG, DeltaBeta_PrimaryCellLine = DeltaBeta_X,DeltaBeta_NormalTissue = DeltaBeta_Y,Class) %>% 
             arrange(desc(Class))
df_in_csv$Class %>% table
write_csv(df_in_csv, path = "XYScatter/ScatterPlot_PrimaryCL_NormTiss.csv")

```



```{r tumor_testing}
pheno_in              <- pData(norm_data_f) %>% as.data.frame 
design                <- model.matrix(~0 + Tiss_Conv2, data = pheno_in)
block_in              <- pheno_in$sample
colnames(design)      <- colnames(design) %>% gsub("Tiss_Conv2","",.)
cont                  <- c(
                           # "Tumour_Tissue"                  = "hgrade_ttissue_BS-hgrade_ttissue_OXBS",
                           # "Normal_Tissue"                  = "normal_ntissue_BS-normal_ntissue_OXBS",
                           "Primary_CellLine"               = "normal_pline_BS-normal_pline_OXBS",
                           # "Whole_Blood"                    = "normal_Blood_BS-normal_Blood_OXBS",
                           # "Brain_Tissue"                   = "normal_Brain_BS-normal_Brain_OXBS",
                           "HighGrade_Established_CellLine" = "hgrade_estline_BS-hgrade_estline_OXBS",
                           "LowGrade_Established_CellLine"  = "lgrade_estline_BS-lgrade_estline_OXBS")
cont_mat             <- makeContrasts(contrasts = cont, levels = colnames(design))
colnames(cont_mat)   <- names(cont)
cont_vec             <- cont %>% gsub("-","_Vs_",.)
mVal                  <- getM(norm_data_f)
bVal                  <- getBeta(norm_data_f)
Corr                 <- duplicateCorrelation(getM(norm_data_f), design, block = block_in)
fitm                 <- lmFit(getM(norm_data_f),    design, block = block_in, correlation = Corr$cor) %>% contrasts.fit(cont_mat)    %>% eBayes
fitb                 <- lmFit(getBeta(norm_data_f), design, block = block_in, correlation = Corr$cor) %>% contrasts.fit(cont_mat) %>% eBayes

fitm                 <- lmFit(getM(norm_data_f),    design) %>% contrasts.fit(cont_mat) %>% eBayes
fitb                 <- lmFit(getBeta(norm_data_f), design) %>% contrasts.fit(cont_mat) %>% eBayes

contrasts            <- colnames(cont_mat)
pVal                 <- 0.05
fc                   <- 0.1

# ttm                  <- topTable(fitm, coef = "TEST1", number = Inf, p.value = pVal, adjust.method = "BH") %>%
#                         as.data.frame %>% add_rownames("CpG") %>% dplyr::select(CpG,`adj.P.Val`)
# ttb                  <- topTable(fitb, coef = "TEST1", number = Inf) %>% as.data.frame %>% add_rownames("CpG") %>%
#                         dplyr::select(CpG,Delta_Beta = logFC)
# list_A               <- ttm %>% left_join(ttb)  %>% left_join(annotation, by = c("CpG" = "Name")) %>%
#                         filter(Delta_Beta > fc) %>% arrange(desc(abs(Delta_Beta)))
# 
# ttm                  <- topTable(fitm, coef = "TEST2", number = Inf, p.value = pVal, adjust.method = "BH") %>%
#                         as.data.frame %>% add_rownames("CpG") %>% dplyr::select(CpG,`adj.P.Val`)
# ttb                  <- topTable(fitb, coef = "TEST2", number = Inf) %>% as.data.frame %>% add_rownames("CpG") %>%
#                         dplyr::select(CpG,Delta_Beta = logFC)
# list_B               <- ttm %>% left_join(ttb)  %>% left_join(annotation, by = c("CpG" = "Name")) %>%
#                         filter(Delta_Beta > fc) %>% arrange(desc(abs(Delta_Beta)))
# 
# list_Int             <- data.frame(CpG = intersect(list_A$CpG, list_B$CpG)) %>% 
#                         left_join(annotation, by = c("CpG" = "Name"))
# 
# write_csv(list_A,   path = "Tumor_BS-OXBS_noBC6.csv")
# write_csv(list_B,   path = "Normal_BS-OXBS_noBC6.csv")
# write_tsv(list_Int, path = "Common_HMc_noBC6.tsv")
# 
# plot(1:nrow(ttb), sort(ttb$Delta_Beta))
# gg <- ggplot(list_A, aes(Delta_Beta, -log10(`adj.P.Val`))) +
#       geom_point() +
#       geom_hline(yintercept = -log10(0.05)) +
#       geom_vline(xintercept = c(0.1,-0.1)) +
#       scale_x_continuous(limits = c(-1, 1)) +
#       theme_bw()
# print(gg)
# list_A
# list_B
# list_Int
# # 
# cpg_list        <- list_A$CpG[1:10]
# 
# for(i in cpg_list) {
#   cpg_in          <- i
#   df              <- bVal[cpg_in,] %>%
#                      as.data.frame %>%
#                      add_rownames("sentrixID") %>%
#                      left_join(pheno_in) %>%
#                      filter(grepl("hgrade_ttissue|normal_ntissue", Tiss_Conv2))
#   colnames(df)[2] <- "Beta"
#   gg              <- ggplot(df, aes(Tiss_Conv2,Beta, colour = tissue_type)) +
#                      geom_point() + geom_line(aes(group = sample)) +
#                      theme_bw() +
#                      # geom_text(aes(label = Sample_Name)) +
#                      ggtitle(i) +
#                      stat_summary(fun.y = "mean", colour = "red", size = 2, geom = "point") +
#                      ylim(c(0,1)) + theme(axis.text.x = element_text(angle = 90, hjust = 1))
#   print(gg)
#   # png(paste0("../Analysis/Healthy_T_Cells/","ExpressionPlot",i,".png"),
#   #     width  = 6.98,height = 6.98,
#   #     units  = "in",res    = 600)
#   # print(gg); dev.off()
# }


# ttm                  <- topTable(fitm, coef = "TEST2", number = Inf, p.value = pVal, adjust.method = "BH") %>%
#                         as.data.frame %>% add_rownames("CpG") %>% dplyr::select(CpG,`adj.P.Val`)
# ttb                  <- topTable(fitb, coef = "TEST2", number = Inf) %>% as.data.frame %>% add_rownames("CpG") %>%
#                         dplyr::select(CpG,Delta_Beta = logFC)
# list_B               <- ttm %>% left_join(ttb)  %>% left_join(annotation, by = c("CpG" = "Name")) %>%
#                         filter(Delta_Beta < fc) %>% arrange(desc(abs(Delta_Beta)))
# 
# list_C               <- intersect(list_A$CpG, list_B$CpG)
# 
# norm_data_tmp        <- norm_data_f[match(list_C,rownames(norm_data_f)),]
# design               <- model.matrix(~0 + Tiss_Conv2, data = pheno_in)
# colnames(design)     <- colnames(design) %>% gsub("Tiss_Conv2","",.)
# cont                 <- c("TEST1"   = "(hgrade_ttissue_BS - hgrade_ttissue_OXBS) - (normal_ntissue_BS - normal_ntissue_OXBS)")
# cont_mat             <- makeContrasts(contrasts = cont, levels = colnames(design))
# colnames(cont_mat)   <- names(cont)
# fitm                 <- lmFit(getM(norm_data_tmp), design) %>% contrasts.fit(cont_mat) %>% eBayes
# fitb                 <- lmFit(getBeta(norm_data_tmp), design) %>% contrasts.fit(cont_mat) %>% eBayes
# 
# ttm                  <- topTable(fitm, coef = "TEST1", number = Inf, p.value = 0.05, adjust.method = "BH") %>%
#                         as.data.frame %>% add_rownames("CpG") %>% dplyr::select(CpG,`adj.P.Val`)
# ttb                  <- topTable(fitb, coef = "TEST1", number = Inf) %>% as.data.frame %>% add_rownames("CpG") %>%
#                         dplyr::select(CpG,Delta_Beta = logFC)
# output               <- ttm %>% left_join(ttb)  %>% left_join(annotation, by = c("CpG" = "Name")) %>%
#                         filter(abs(Delta_Beta) > fc) %>% arrange(desc(abs(Delta_Beta)))
# output
```

```{r model_fit_paired, cache=FALSE, message=FALSE, warning=FALSE, error=FALSE, results='hide'}
pheno_in              <- pData(norm_data_f) %>% 
                         as.data.frame 

mVal                  <- getM(norm_data_f)
bVal                  <- getBeta(norm_data_f)
design                <- model.matrix(~0 + Tiss_Conv2, data = pheno_in)
Var_Levels            <- pheno_in$Tiss_Conv %>% factor %>% levels
block_in              <- pheno_in$sample

colnames(design)      <- colnames(design) %>% gsub("Tiss_Conv2","",.)
cont                  <- c("TEST1"   = "normal_pline_BS-normal_pline_OXBS",
                           "TEST2"   = "normal_ntissue_BS-normal_ntissue_OXBS")

cont_mat             <- makeContrasts(contrasts = cont, levels = colnames(design))
colnames(cont_mat)   <- names(cont)
cont_vec             <- cont %>% gsub("-","_Vs_",.)
# fitm                 <- lmFit(mVal, design) %>% contrasts.fit(cont_mat) %>% eBayes
# fitb                 <- lmFit(bVal, design) %>% contrasts.fit(cont_mat) %>% eBayes

# Paired Model
Corr <- duplicateCorrelation(mVal, design, block = block_in)
fitm                 <- lmFit(mVal, design, block = block_in, correlation = Corr$cor) %>% contrasts.fit(cont_mat) %>% eBayes
fitb                 <- lmFit(bVal, design, block = block_in, correlation = Corr$cor) %>% contrasts.fit(cont_mat) %>% eBayes
#

contrasts            <- colnames(cont_mat)
pVal                 <- 0.05
fc                   <- 0.1
```

```{r model_fit_group, cache=FALSE, message=FALSE, warning=FALSE, error=FALSE, results='hide'}
pheno_in              <- pData(norm_data_f) %>% 
                         as.data.frame 

mVal                  <- getM(norm_data_f)
bVal                  <- getBeta(norm_data_f)
design                <- model.matrix(~0 + Tiss_Conv2, data = pheno_in)
Var_Levels            <- pheno_in$Tiss_Conv %>% factor %>% levels

colnames(design)      <- colnames(design) %>% gsub("Tiss_Conv2","",.)
cont                  <- c("TEST1"   = "hgrade_ttissue_OXBS-normal_ntissue_OXBS",
                           "TEST2"   = "hgrade_ttissue_BS-normal_ntissue_BS")

cont_mat             <- makeContrasts(contrasts = cont, levels = colnames(design))
colnames(cont_mat)   <- names(cont)
cont_vec             <- cont %>% gsub("-","_Vs_",.)
fitm                 <- lmFit(mVal, design) %>% contrasts.fit(cont_mat) %>% eBayes
fitb                 <- lmFit(bVal, design) %>% contrasts.fit(cont_mat) %>% eBayes
contrasts            <- colnames(cont_mat)
pVal                 <- 0.01
fc                   <- 0.1
```


```{r Results, cache=FALSE, message=FALSE, warning=FALSE, error=FALSE, results='hide'}
tt_out  <- list()
ttu_out <- list()
for(i in names(cont_vec)) {
  ttm                  <- topTable(fitm, coef = i, number = Inf, p.value = pVal, adjust.method = "BH") %>% 
                          as.data.frame %>% add_rownames("CpG")
  ttum                 <- topTable(fitm, coef = i, number = Inf, adjust.method = "BH") %>% 
                          as.data.frame %>% add_rownames("CpG")
  if(nrow(ttm) > 0) {
    ttm                  <- ttm %>% dplyr::select(CpG,`adj.P.Val`)
    ttb                  <- topTable(fitb, coef = i, number = Inf) %>% 
                            as.data.frame %>% add_rownames("CpG") %>% dplyr::select(CpG,Delta_Beta = logFC)
    tt                   <- ttm %>% left_join(ttb)  %>% 
                                    left_join(annotation, by = c("CpG" = "Name")) %>% 
                                    filter(abs(Delta_Beta) > fc) %>% arrange(desc(abs(Delta_Beta)))
    
    ttum                 <- ttum %>% dplyr::select(CpG,`adj.P.Val`)
    ttub                 <- topTable(fitb, coef = i, number = Inf) %>% 
                            as.data.frame %>% add_rownames("CpG") %>% dplyr::select(CpG,Delta_Beta = logFC)
    ttu                  <- ttum %>% left_join(ttub)  %>% 
                                     left_join(annotation, by = c("CpG" = "Name"))
    
    tt %>% print
    tt_out[[i]]          <- tt
    ttu_out[[i]]         <- ttu
    # write_tsv(tt, path = paste0("./BS_-_OXBS_DMTesting/DiffMethyl_BS_OXBS_",i,".txt"))
  } else {
    paste0(i, ": No DM Results") %>% print
  }
}

# tt_out$RA_HC %>% filter(UCSC_RefGene_Name %in% c("MAGI2","DUSP22","AGPAT9") )
# foo <- read_tsv("BS_-_OXBS_DMTesting/DiffMethyl_BS_OXBS_Primary_CellLine.txt") %>% 
#        as.data.frame
# foo$adj.P.Val %>% range
```

```{r visualise_result}
cpg_list        <- tt_out$TEST1$CpG[1]

for(i in cpg_list) {
  cpg_in          <- i
  df              <- bVal[cpg_in,] %>% 
                     as.data.frame %>% 
                     add_rownames("sentrixID") %>% 
                     left_join(pheno_in)
  colnames(df)[2] <- "Beta" 
  gg              <- ggplot(df, aes(Tiss_Conv2,Beta, colour = tissue_type)) +
                     geom_point() + geom_line(aes(group = sample)) +
                     theme_bw() +
                     # geom_text(aes(label = Sample_Name)) +
                     ggtitle(i) +
                     stat_summary(fun.y = "mean", colour = "red", size = 2, geom = "point") +
                     ylim(c(0,1)) + theme(axis.text.x = element_text(angle = 90, hjust = 1))
  print(gg)
  # png(paste0("../Analysis/Healthy_T_Cells/","ExpressionPlot",i,".png"),
  #     width  = 6.98,height = 6.98,
  #     units  = "in",res    = 600)
  # print(gg); dev.off()
}
```



```{r upset_and_barchart}
library(UpSetR)

df_upset          <- data.frame(CpG = "")
for(i in names(tt_out)) {
  tmpdf              <- data.frame(CpG = tt_out[[i]]$CpG, DM = 1)
  colnames(tmpdf)[2] <- paste0("DM_", i)
  
  df_upset           <- df_upset %>% 
                        full_join(tmpdf)
  
  # Write out sets
  # oxbs_bs_intersection <- data.frame(Gene_Symbol = intersect(unique(dmr_to_gene_out[["OXBS"]][[i]]$Gene),
  #                                                          unique(dmr_to_gene_out[["BS"]][[i]]$Gene)))
  # oxbs_only            <- data.frame(Gene_Symbol = setdiff(unique(dmr_to_gene_out[["OXBS"]][[i]]$Gene),
  #                                                          unique(dmr_to_gene_out[["BS"]][[i]]$Gene)))
  # bs_only              <- data.frame(Gene_Symbol = setdiff(unique(dmr_to_gene_out[["BS"]][[i]]$Gene),
  #                                                          unique(dmr_to_gene_out[["OXBS"]][[i]]$Gene)))
  # write_csv(oxbs_bs_intersection, path = paste0("Sets/OXBS_BS_Intersect_",i,".csv"))
  # write_csv(oxbs_only, path = paste0("Sets/OXBS_Only_",i,".csv"))
  # write_csv(bs_only, path = paste0("Sets/BS_Only_",i,".csv"))
}
df_upset[is.na(df_upset)] <- 0


png("Sets/UpSetPlot_CellTypes_BS-OXBS.png",
      width  = 13,height = 7.5,
      units  = "in",res    = 600)
upset(df_upset, nsets = 7); dev.off()


# for(i in names(tt_out)) {
#   df_upset_tmp <- df_upset %>% dplyr::select(CpG, ends_with(i))
#   png(paste0("Sets/UpSetPlot_CellTypes_BS-OXBS_",i,".png"),
#       width  = 13,height = 7.5,
#       units  = "in",res    = 600)
#   upset(df_upset_tmp, nsets = 2); dev.off()
# }

df_bar <- c()
for(i in names(tt_out)) {
  tmp_df     <- df_upset %>% dplyr::select(ends_with(i))
  tmp_sum    <- sum(tmp_df[[1]])
  tmp_df_out <- data.frame(Tissue = i, 
                           Class  = c("Sum"),
                           Counts = c(tmp_sum))
  df_bar     <- rbind(df_bar, tmp_df_out)
}
df_bar_in <- df_bar
gg        <- ggplot(df_bar_in, aes(Tissue,Counts)) +
             geom_bar(stat = "identity", position = "dodge") + 
             theme_bw() + 
             geom_text(aes(label=Counts), vjust=0) +
             theme(axis.text.x=element_text(angle = -90, hjust = 0)) +
             ggtitle("The Number of 5HmC Sites found in each Tissue Group")

png("Sets/BarChart.png",
      width  = 7.5,height = 7.5,
      units  = "in",res    = 600)
print(gg); dev.off()


```




```{r DMR_Detection}
# design_DMR                <- model.matrix(~sample + Tiss_Conv, data = pheno_in)
# colnames(design_DMR)      <- colnames(design_DMR) %>% gsub("Tiss_Conv","",.)

DMR_List_Out <- list()
for(i in colnames(cont_mat)) {
  myannotation       <- cpg.annotate(datatype      = "array",
                                     what          = "M",
                                     arraytype     = "450K",
                                     object        = mVal, 
                                     analysis.type = "differential", 
                                     design        = design, 
                                     coef          = i,
                                     contrasts     = T, 
                                     cont.matrix   = cont_mat,
                                     fdr           = 0.01)
  dmrcoutput         <- dmrcate(myannotation) 
                      
  dmrresults         <- dmrcoutput$results %>% 
                        as.data.frame %>% 
                        mutate(betaAfc = abs(meanbetafc)) %>% 
                        filter(betaAfc > 0.1) %>% 
                        arrange(desc(no.cpgs))
  
  dmr_genes_tmp <- sapply(dmrresults$coord, function(x) {
    dmr_coords           <- x
    dmr_chr              <- strsplit(dmr_coords, ':')[[1]][1]
    dmr_start            <- as.numeric(strsplit(dmr_coords, '-|:')[[1]][2])
    dmr_end              <- as.numeric(strsplit(dmr_coords, '-')[[1]][2])
    dmr_anno             <- annotation %>% 
                            filter(chr == dmr_chr,
                                   pos >= (dmr_start),
                                   pos <= (dmr_end))
    dmr_genes            <- dmr_anno$UCSC_RefGene_Name %>% unique %>% .[. != ""]
    dmr_enhancer         <- dmr_anno$Enhancer %>% unique %>% .[. != ""]
    if(length(dmr_genes) == 0 & length(dmr_enhancer) == 0) {
      return(c("",""))
    } else {
      dmr_genes_tmp        <- sapply(dmr_genes, function(y) {
        gene_split_tmp       <- strsplit(y,";")[[1]] %>% unique
        return(gene_split_tmp)
      })
      dmr_enhancers_tmp    <- sapply(dmr_enhancer, function(y) {
        enhancer_split_tmp   <- strsplit(y,";")[[1]] %>% unique
        return(enhancer_split_tmp)
      })
      
      dmr_genes            <- dmr_genes_tmp %>% unlist %>% unique %>% paste(., collapse = ";")
      dmr_enhancers        <- dmr_enhancers_tmp %>% unlist %>% unique %>% paste(., collapse = ";")
      return(c(dmr_genes,dmr_enhancers))
      # return(dmr_genes)
    }
  })
  gene_enhancer_tmp <- dmr_genes_tmp %>% unlist
  
  
  dmrresults$Genes      <- gene_enhancer_tmp[seq(1,length(gene_enhancer_tmp),2)]
  dmrresults$Enhancers  <- gene_enhancer_tmp[seq(2,length(gene_enhancer_tmp),2)]
  DMR_List_Out[[i]] <- dmrresults
  print(dmrresults)
}
  
write_csv(DMR_List_Out[[1]], path = "DMR_Results_OXBS_2Jun_NoBc6 2018.csv")
write_csv(DMR_List_Out[[2]], path = "DMR_Results_BS_2Jun_NoBc6 2018.csv")

# write_tsv(dmrresults, path = "DMR_Results_noBC6_1.tsv")
# dmrresults
# write_csv(dmrresults, path = paste0("../Analysis/Healthy_T_Cells/DMR_", names(cont_vec)[i], ".csv"))
# save(dmrresults, bVal, pheno_in, file = "DMR_Temp_Data.RData", compress = T)
```

```{r plot_dmr, warning=FALSE, message=FALSE}
for(i in 1:10) {
  dmr_offset           <- 1000
  dmr_coords           <- dmrresults$coord[i]
  dmr_chr              <- strsplit(dmr_coords, ':')[[1]][1]
  dmr_start            <- as.numeric(strsplit(dmr_coords, '-|:')[[1]][2])
  dmr_end              <- as.numeric(strsplit(dmr_coords, '-')[[1]][2])
  dmr_anno             <- annotation %>% 
                          filter(chr == dmr_chr,
                                 pos >= (dmr_start - dmr_offset),
                                 pos <= (dmr_end + dmr_offset))
  df                   <- bVal %>% as.data.frame %>% 
                          add_rownames("CpG") %>% 
                          filter(CpG %in% dmr_anno$Name) %>% 
                          melt %>% dplyr::select(CpG,sentrixID=variable,Beta=value) %>% 
                          left_join(as.data.frame(pData(norm_data_f))) %>% 
                          left_join(annotation, by = c("CpG"="Name"))
  gg                   <- ggplot(df,aes(pos,Beta,colour=Tiss_Conv)) +
                          scale_colour_hue(l = 50) + geom_point() +
                          geom_vline(xintercept = c(dmr_start,dmr_end),colour ='red') +
                          theme_bw() +
                          stat_summary(fun.y = mean, geom = "line", aes(colour = Tiss_Conv, 
                                       group = Tiss_Conv)) +
                          ylim(c(0, 1)) + 
                          facet_grid(Tiss_Conv~.)+
                          ggtitle(paste0("DMR Found: ", dmr_coords))
  print(gg)
  
  # png(paste0("./","DMRPlot",i,".png"),
  #     width  = 6.98,height = 6.98,
  #     units  = "in",res    = 600)
  # print(gg); dev.off()
}
```


```{r DMR_Prep}
gtf_path      <- "genome.gtf"
transcriptdb  <- makeTxDbFromGFF(file     = gtf_path,
                                 format   = 'gtf',
                                 organism = 'Homo sapiens')
ensembl       <- useMart("ENSEMBL_MART_ENSEMBL",
                         dataset="hsapiens_gene_ensembl",
                         host="grch37.ensembl.org",
                         path="/biomart/martservice")
```



```{r plot_dmr2, warning=FALSE, message=FALSE}
# for(i in 1:2) {
  # i=156
bVal <- getBeta(norm_data_f)
  dmr_offset           <- 1000
  # dmr_coords           <- dmrresults$coord[i]
  dmr_coords           <- "chr2:50574196-50575098"
  dmr_chr              <- strsplit(dmr_coords, ':')[[1]][1]
  dmr_start            <- as.numeric(strsplit(dmr_coords, '-|:')[[1]][2])
  dmr_end              <- as.numeric(strsplit(dmr_coords, '-')[[1]][2])
  dmr_anno             <- annotation %>% 
                          filter(chr == dmr_chr,
                                 pos >= dmr_start,
                                 pos <= dmr_end)
  dmr_genes            <- dmr_anno$UCSC_RefGene_Name %>% unique %>% .[. != ""]
  if(length(dmr_genes) > 0) {
    dmr_genes_tmp <- sapply(dmr_genes, function(x) {
      gene_split_tmp <- strsplit(x,";")[[1]] %>% unique
      return(gene_split_tmp)
    })
    dmr_genes <- dmr_genes_tmp %>% unique %>% paste(., collapse = ";")
  } else {
    dmr_genes <- ""
  }                       

  bump_pos_limit       <- c(dmr_start  - dmr_offset,
                            dmr_end    + dmr_offset)
  
  df                   <- bVal %>% as.data.frame %>% 
                          add_rownames("CpG") %>% 
                          filter(CpG %in% dmr_anno$Name) %>% 
                          melt %>% dplyr::select(CpG,sentrixID=variable,Beta=value) %>% 
                          left_join(as.data.frame(pData(norm_data_f))) %>% 
                          left_join(annotation, by = c("CpG"="Name"))
  
  # Uncomment for OXBS Only
  df <- df %>% filter(conversion == "OXBS")
  #
  
  # print(gg)
  
  geneID_tmp           <- transcriptsByOverlaps(transcriptdb, 
                                                GRanges(seqnames = as.numeric(gsub("chr","",dmr_chr)), 
                                                        IRanges(dmr_start - dmr_offset, dmr_end + dmr_offset)))
  geneID_in            <- subsetByOverlaps(geneID_tmp,
                                           GRanges(seqnames = as.numeric(gsub("chr","",dmr_chr)), 
                                                   IRanges(dmr_start,dmr_end)))
  if(length(geneID_in) > 0) {
    annotation.in         <- getBM(attributes = c("ensembl_gene_id",
                                                  "external_gene_name"),
                                   filters    = "ensembl_transcript_id",
                                   values     = as.vector(geneID_in@elementMetadata@listData$tx_name[1]),
                                   mart       = ensembl)
    annotation.in         <- annotation.in$external_gene_name
    annotation.in         <- paste0("Close to ", paste(annotation.in, collapse = ";"))
  } else {
    annotation.in <- "No Data"
  }
  
  gg                   <- ggplot(df,aes(pos,Beta,colour=Tiss_Conv2)) +
                          scale_colour_hue(l = 50) + 
                          # geom_point() +
                          geom_vline(xintercept = c(dmr_start,dmr_end),colour ='red') +
                          theme_bw() + xlab("") +
                          stat_summary(fun.y = mean, geom = "line", aes(colour = Tiss_Conv2, 
                                       group = Tiss_Conv2)) +
                          stat_summary(fun.y = mean, geom="point", aes(colour = Tiss_Conv2, 
                                       group = Tiss_Conv2)) +
                          ylim(c(0, 1)) + 
                          ggtitle(paste0("DMR Found: ", dmr_coords, "\n", annotation.in))
  
  # if(length(geneID_tmp) == 0) {
  #   p1_empty <- data.frame(x=bump_pos_limit, y=c(4,5,6,7))
  #   p1       <- ggplot(p1_empty, aes(x=x, y=y)) +
  #               geom_blank() + theme_bw() +
  #               scale_x_continuous(limits = bump_pos_limit) +
  #               geom_vline(xintercept     = dmr_start,
  #                          colour         = "red")  +
  #               geom_vline(xintercept     = dmr_end,
  #                          colour         = "red") +
  #               labs(x="",y="") +
  #               theme(axis.text.y=element_blank())
  # } else {
  #   trans    <- crunch(transcriptdb, which=geneID_tmp)
  #   gr1      <- split(trans,trans$tx_name)
  #   p1       <- ggbio::autoplot(gr1) + theme_bw() +
  #               geom_vline(xintercept     = dmr_start,
  #                          colour         = "red") +
  #               geom_vline(xintercept     = dmr_end,
  #                          colour         = "red")
  #   p1       <- p1@ggplot
  # }

  ##Plot2 - Strip Plot of Site Type
  gg2 <- ggplot(df, aes(x = pos, y = Relation_to_Island)) +
         theme_bw() + geom_point(size = 3) +
         ylab("") + xlab("") +
         theme(legend.title         = element_blank(),
               legend.position      = "bottom",
               legend.margin        = unit(-0.6,"cm")) +
         scale_x_continuous(limits = bump_pos_limit) +
         geom_vline(xintercept      = c(dmr_start,dmr_end),
                    colour          = "red")
  # ##

  # Plot2 - Karyogram
  # data(hg19IdeogramCyto, package = "biovizBase")
  # getOption("biovizBase")$cytobandColor
  # hg19IdeogramCyto <- hg19IdeogramCyto[hg19IdeogramCyto@seqnames == dmr_chr,]
  # dmr_rng          <- GRanges(seqnames = dmr_chr, IRanges(dmr_start, dmr_end))
  # p                <- ggplot(hg19IdeogramCyto) + layout_karyogram(cytoband = T) +
  #                     layout_karyogram(dmr_rng, geom = "rect", ylim = c(11, 21),
  #                                      color = "red", fill = "red")
  # p                <- p@ggplot
  # p                <- p + theme(legend.position="none") + xlab("")
  ##
  
  ggOut <- plot_grid(gg,gg2,
                   align       = 'v', 
                   labels      = c('A', 'B'),
                   nrow        = 2,
                   rel_heights = c(2, 1))
  print(ggOut)

  # png(paste0("./Analysis_Final_Structure/Results/Methylation/DMR/","DMR_Conversion MAGI2_BS.png"),
  #     width  = 6.98,height = 6.98,
  #     units  = "in",res    = 600)
  # print(ggOut); dev.off()

```



```{r Make_Bed_Tracks_Background}
# tt_out - List of DM results / annotation - background universe set
anno_in    <- annotation
out        <- data.frame(chr = anno_in$chr) %>% 
              mutate(start  = anno_in$pos - 1, end = as.numeric(anno_in$pos), 
                     name   = paste0(anno_in$Name),
                     score  = as.numeric(0),
                     strand = as.vector(anno_in$strand),
                     box_s  = as.numeric(`start`),
                     box_e  = as.numeric(`end`),
                     colour = "0,0,0")
  out$box_s <- format(out$box_s, scientific = F)
  out$end <- format(out$end, scientific = F)
  out$start <- format(out$start, scientific = F)
  out$box_e <- format(out$box_e, scientific = F)

write_tsv(out, path = "Background_noBC6.bed", col_names = F)
# exe <- paste0("sed -i '1s/^/track name=\"", 
#               "Background", "\" description=\"",
#               "Background", "\" visibility=3 ",
#               "itemRgb=\"On\"",
#               "\\n/' ", 
#               "Background.bed")
# system(exe)


for(i in 1:length(tt_out)) {
  tt         <- tt_out[[i]]
  tt_name    <- names(tt_out)[i]
   
  anno_in    <- annotation %>% dplyr::slice(match(tt$CpG,Name))
  out        <- data.frame(chr = anno_in$chr) %>% 
                mutate(start  = anno_in$pos - 1, end = as.numeric(anno_in$pos), 
                       name   = anno_in$Name,
                       score  = 0,
                       strand = as.vector(anno_in$strand),
                       box_s  = as.numeric(`start`),
                       box_e  = as.numeric(`end`),
                       colour = "0,0,0")
  options(scipen=999)
  out$box_s <- format(out$box_s, scientific = F)
  out$end <- format(out$end, scientific = F)
  out$start <- format(out$start, scientific = F)
  out$box_e <- format(out$box_e, scientific = F)
  
  write_tsv(out, path = paste0("DM", tt_name, ".bed"), col_names = F)
}


out %>% filter(box_s == 98246e3)
```


```{r TCGA_data}
# projects <- TCGAbiolinks:::getGDCprojects()$project_id
# projects <- projects[grepl('^TCGA-BLCA',projects,perl=T)]
# match.file.cases.all <- NULL
# for(proj in projects){
#     print(proj)
#     query <- GDCquery(project = proj,
#                       data.category = "Raw microarray data",
#                       data.type = "Raw intensities", 
#                       experimental.strategy = "Methylation array", 
#                       legacy = TRUE,
#                       file.type = ".idat",
#                       platform = "Illumina Human Methylation 450")
#     match.file.cases <- getResults(query,cols=c("cases","file_name"))
#     match.file.cases$project <- proj
#     match.file.cases.all <- rbind(match.file.cases.all,match.file.cases)
#     tryCatch(GDCdownload(query, method = "api",chunks.per.download = 20),
#              error = function(e) GDCdownload(query, method = "client"))
# }
# # This will create a map between idat file name, cases (barcode) and project
# readr::write_tsv(match.file.cases.all, path =  "idat_filename_case.txt")
# # code to move all files to local folder
# for(file in dir(".",pattern = ".idat", recursive = T)){
#     TCGAbiolinks::move(file,basename(file))
# }
legacyPipeline <- function(project, data.category, platform){
    query <- GDCquery(project = project,
                      data.category = data.category,
                      platform = platform,
                      legacy = TRUE)
    cases <- query$results[[1]]$cases[1:2]
    query <- GDCquery(project = project,
                      data.category = data.category,
                      platform = platform,
                      legacy = TRUE,
                      barcode = cases)
    GDCdownload(query)
    data <- GDCprepare(query)
    return(data)
}
data <- legacyPipeline("TCGA-BLCA","DNA methylation","Illumina Human Methylation 450")
```





```{r QC}
summary <- shinySummarize(raw_data)
runShinyMethyl(summary)
```

```{r beta_value}
betas = getBeta(norm_data)
betas
betas_output <- betas %>% as.data.frame %>% add_rownames("CpG")
write_csv(betas_output, path = "normalised_betas_noBC6.csv")
```

```{r cellcounts}
cellcounts <- minfi::estimateCellCounts(raw_data)

df_cc <- cellcounts %>% 
         melt %>% 
         left_join(pheno_in, by = c("Var1"= "sentrixID")) 
          
gg    <- ggplot(df_cc, aes(Var2,value)) + 
         geom_point() + 
         theme_bw() +
         facet_grid(.~tissue_type, scales="free") +
         geom_text(aes(label = sample))
print(gg)
```